src/core/reactor_backend.cc

```diff
@@ -1316,11 +1331,140 @@ class reactor_backend_uring final : public reactor_backend {
         auto sqe = get_sqe();
         ::io_uring_prep_poll_add(sqe, fd.fd.get(), events);
         auto ufd = static_cast<uring_pollable_fd_state*>(&fd);
+#ifdef USE_PASSTHRU
+        auto arg = new passthru_arg;
+        arg->completion = ufd->get_desc(events);
+        ::io_uring_sqe_set_data(sqe, arg);
+#else
         ::io_uring_sqe_set_data(sqe, static_cast<kernel_completion*>(ufd->get_desc(events)));
+#endif
         _has_pending_submissions = true;
         return ufd->get_completion_future(events);
     }
 
+#ifdef USE_PASSTHRU
+    struct block_uring_cmd {
+        __u32   ioctl_cmd;
+        __u32   unused1;
+        __u64   unused2[4];
+    };
+
+    struct nvme_user_io_t {
+        uint32_t cdw00_09[10];
+        uint64_t slba;
+        uint32_t nlb  : 16;
+        uint32_t rsvd :  4;
+        uint32_t dtype  :  4;
+        uint32_t rsvd2  :  2;
+        uint32_t prinfo :  4;
+        uint32_t fua  :  1;
+        uint32_t lr :  1;
+        uint32_t cdw13_15[3];
+    };
+
+    struct nvme_io_command {
+        union {
+#ifdef NVME_IOCTL_IO64_CMD
+            nvme_passthru_cmd64 common;
+#else
+            nvme_passthru_cmd common;
+#endif
+            nvme_user_io_t rw;
+            uint32_t raw[16];
+        };
+
+        enum class opcode {
+            FLUSH = 0x0,
+            WRITE = 0x1,
+            READ = 0x2,
+            APPEND = 0x7D,
+        };
+    };
+    struct passthru_arg;
+    class passthru_completion final : public kernel_completion {
+        public:
+            io_completion* orig_completion = nullptr;
+            passthru_arg* arg = nullptr;
+            size_t block_size = -1;
+
+            void complete_with(ssize_t res) override {
+                if (res >= 0) {
+                    complete(res);
+                    return;
+                }
+                ++engine()._io_stats.aio_errors;
+                try {
+                    throw_kernel_error(res);
+                } catch (...) {
+                    set_exception(std::current_exception());
+                }
+            }
+
+            void complete(size_t res) noexcept {
+                if (res == 0) {
+                    orig_completion->set_additional_result(arg->cmd.common.result * block_size);
+                    orig_completion->complete_with(arg->cmd.common.data_len);
+                } else {
+                    // TODO pass proper error code
+                    orig_completion->complete_with(-1);
+                }
+                delete this;
+            }
+
+            void set_exception(std::exception_ptr eptr) noexcept {
+                orig_completion->set_exception(eptr);
+            }
+
+            ~passthru_completion() {}
+    };
+
+    struct passthru_arg {
+        nvme_io_command cmd;
+        kernel_completion* completion;
+    };
+
+    void prep_passthru_append(struct io_uring_sqe *sqe, int fd, void *buf, unsigned nbytes, off_t offset, uint32_t nsid, size_t block_size, io_completion* io_completion) {
```

Comment:
This helper is not exception-safe. If `auto completion = new passthru_completion;` below throws the `auto arg = new passthru_arg;` leaks